edgedb/gelalchemy/models.py

```diff
@@ -0,0 +1,390 @@
+    # is allowed).
+
+    # Clear module
+    name = name.rsplit('::', 1)[-1]
+    # TODO: once modules are supported this function may need to do a bit more
+
+    return name
+
+
+class ModelGenerator(object):
+    INDENT = ' ' * 4
+
+    def __init__(self, *, out=None):
+        if out is not None:
+            self.out = out
+        else:
+            self.out = sys.stdout
+
+        self._indet_level = 0
+
+    def indent(self):
+        self._indet_level += 1
+
+    def dedent(self):
+        if self._indet_level > 0:
+            self._indet_level -= 1
+
+    def reset_indent(self):
+        self._indet_level -= 0
+
+    def write(self, text=''):
+        print(
+            textwrap.indent(text, prefix=self.INDENT*self._indet_level),
+            file=self.out,
+        )
+
+    def render_models(self, spec):
+        self.write(MODELS_STUB)
+
+        for rec in spec['link_tables']:
+            self.render_link_table(rec)
+            self.write()
+
+        lobj_map = {}
+        for lobj in spec['link_objects']:
+            lobj_map[lobj['name']] = lobj
+            self.render_link_object(lobj)
+            self.write()
+
+        pobj_map = {}
+        for pobj in spec['prop_objects']:
+            pobj_map[pobj['name']] = pobj
+            self.render_prop_object(pobj)
+            self.write()
+
+        for rec in spec['object_types']:
+            self.render_type(rec, lobj_map, pobj_map)
+            self.write()
+
+    def render_link_table(self, spec):
+        self.write(f'{spec["name"]} = Table(')
+        self.indent()
+        self.write(f'{spec["table"]!r},')
+        self.write(f'Base.metadata,')
+        self.write(f'Column("source", ForeignKey({spec["source"]!r})),')
+        self.write(f'Column("target", ForeignKey({spec["target"]!r})),')
+        self.dedent()
+        self.write(f')')
+
+    def render_link_object(self, spec):
+        # assume nice names for now
+        name = get_clean_name(spec['name'])
+        sql_name = spec['table']
+        source_link = sql_name.split('.')[-1]
+
+        self.write(f'class {name}(Base):')
+        self.indent()
+        self.write(f'__tablename__ = {sql_name!r}')
+
+        # No ids for these intermediate objects
+        if spec['links']:
+            self.write()
+            self.write('# Links:')
+
+            for link in spec['links']:
+                lname = link['name']
+                target = get_clean_name(link['target']['name'])
+                self.write(f'{lname}_id: Mapped[uuid.UUID] = mapped_column(')
+                self.indent()
+                self.write(f'{lname!r}, Uuid(), ForeignKey("{target}.id"),')
+                self.write(f'primary_key=True, nullable=False,')
+                self.dedent()
+                self.write(')')
+
+                if lname == 'source':
+                    bklink = source_link
+                else:
+                    bklink = f'backlink_via_{source_link}'
```

Comment:
What is there are multiple incoming links with that name?

Reply:
I don't really have a handler for backlink name collisions. I'm tempted to just start adding numbers to them because adding anything longer to the names (referring to the object the backlink is from) is probably too much for a name.

Reply:
Fixed